Docs: Add Building a CLI Application How to

john0isaac opened the pull request:
fix #1947 

## PR Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [-] All new features have been tested
- [-] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

freakboy3742 reviewed (changes requested):
Thanks - this is a good first draft. The underlying example is a good target, and you've covered most of the major points that we need to hit. 

The major issue that needs to be resolved is whether this should be a tutorial or a HOWTO. I suspect that you've landed on writing this as a HOWTO because of the existing language in the tutorials section leading to the HOWTO page - but that should be read more of an indication that we need to lay a path to *somewhere*, and in the absence of more tutorial pages, HOWTOs are the next step.

We use [diataxis](https://diataxis.fr) as a guide for how we structure our docs. In that framework, a tutorial is a complete step-by-step guide, coming from a clean starting point (or at least a *known* starting point). It's a learning guide, aiding knowledge acquisition. A HOWTO guide, on the other hand, is focused on achieving a specific task. It can assume some background knowledge about how the tool works. It's a more practical guide to completing a specific task. Generally, a HOWTO will be a lot shorter, as it will point out the major steps and landmarks, but won't hold your hand the whole way.

A HOWTO for building a CLI would be *really* short. It wouldn't need to be much more than a bunch of bullet points highlighting the major differences between a GUI and CLI app built with Briefcase.

Therefore, I think I'd lean towards the tutorial in this case. If you've already got a good handle on how Briefcase works, then the process of building a CLI tool shouldn't be *that* unexpected. The only thing that you might not already know about is the handling of command line arguments, and how the app will appear on the command line.

However, even in a tutorial, we can lean heavily on assuming the user has done the existing BeeWare tutorial. I'd suggest structuring this document as the first of a collection of "additional tutorial topics" - similar to how the [end of the BeeWare tutorial](https://docs.beeware.org/en/latest/tutorial/topics/index.html) is structured. Those tutorials assume you've already done the main tutorial (and explicitly say as much), but build on that knowledge with a new example. 

In that context, you don't need to be explicit about *all* the initial setup steps - you should definitely summarise the main steps (create a fresh venv, install Briefcase, create an app with a name of `hello-cli`, select Console as the GUI framework); but that's more by way of refreshing the user's memory about the steps required, not explicit handholding. You can defer to "as you did in step X of the beeware tutorial"-style instructions, rather than giving full explicit console output of *every* step. You only need to be explicit when something *new* is happening - for example, in the packaging step, showing how the installed app would be executed.

I've left a few other comments inline; if you need any clarification on any of this, I'm happy to fill in the gaps.

john0isaac commented on docs/how-to/cli-apps.rst:
```diff
@@ -0,0 +1,668 @@
+Next steps
+----------
+
+So far we have been running the application in developer mode. To distribute the application, you will need to package it for distribution.
```
sure, I will add the packaging steps

john0isaac commented on docs/how-to/cli-apps.rst:
```diff
@@ -0,0 +1,668 @@
+    .. code-block:: console
+
+      (venv) $ briefcase new
+      Let's build a new Briefcase app!
```
Done

john0isaac commented on docs/how-to/cli-apps.rst:
```diff
@@ -0,0 +1,668 @@
+
+* Python 3.8 or later
+* Briefcase
+* A code editor
```
Done

john0isaac commented on docs/how-to/cli-apps.rst:
```diff
@@ -0,0 +1,668 @@
+Overview
+---------
+
+In this tutorial, you'll learn how to build and package a Command Line Interface (CLI) application for distribution on Windows, macOS, and Linux using Briefcase.
```
moved to tutorials

john0isaac commented on docs/how-to/cli-apps.rst:
```diff
@@ -0,0 +1,668 @@
+------------------------
+
+Let's start our first Briefcase CLI project!
+We're going to use the Briefcase ``new`` command to create an application called **Hello World**.
```
used "Hello CLI"

john0isaac commented on docs/how-to/cli-apps.rst:
```diff
@@ -0,0 +1,668 @@
+----------------------
+
+Right now the ``app.py`` file contains a simple ``print`` statement that prints "Hello, World.".
+Let's use ``argparse`` to make it more interesting.
```
I made all the lines shorter than 88 character

john0isaac commented on docs/how-to/cli-apps.rst:
```diff
@@ -0,0 +1,668 @@
+----------------------
+
+Right now the ``app.py`` file contains a simple ``print`` statement that prints "Hello, World.".
+Let's use ``argparse`` to make it more interesting.
```
Done

freakboy3742 commented:
Just to confirm - is this ready for another review, or are you expecting to make more updates?

john0isaac commented:
Not yet, the packaging steps are still not done, when it's done I will re-request your review.
I will do it in the next couple of hours.
Thanks for checking.

john0isaac commented:
@freakboy3742 Sorry for the delay I finally was able to test this on Linux and Windows.
I addressed all of the points you raised and added steps till installing the packaged app and using it in the terminal.

I have a question thought why is the size of the application on macOS +200MB this is huge for a program that only says Hello + Your Name 😆 I think it would be great if we had a way to make it slimmer.

Thank you for your patience!
Looking forward to your review and any other required changes.

freakboy3742 reviewed (changes requested):
Thanks for these updates; however, I think we've had a miscommunication over my previous review notes. In my last review I said:

> In that context, you don't need to be explicit about all the initial setup steps - you should definitely summarise the main steps (create a fresh venv, install Briefcase, create an app with a name of hello-cli, select Console as the GUI framework); but that's more by way of refreshing the user's memory about the steps required, not explicit handholding. You can defer to "as you did in step X of the beeware tutorial"-style instructions, rather than giving full explicit console output of every step. You only need to be explicit when something new is happening - for example, in the packaging step, showing how the installed app would be executed.
 
The bulk of what you've got here is essentially a duplicate of what is contained in the initial BeeWare tutorial. The first ~180 lines of this content could essentially be replaced with "Start a new app as you did in the Beeware tutorial, but select 'console app' rather than 'Toga' as your GUI framework". Similarly for L293 onwards - that can be replace with "You can now build and publish your app as you did in the Beeware tutorial" (ok - maybe a few more words than that.. but not *that* many).

The thing this tutorial can offer is:
1. Explicitly pointing out what is different when you pick "console app"; 
2. Pointing out the `console_app` flag in pyproject.toml as the major thing that is different
3. Highlighting how command line arguments are handled by `briefcase dev` and `briefcase run`
4. Pointing out notable differences in how console apps operate (e.g., outputting to console, rather than to the system log as GUI apps do) 
5. Pointing out notable differences in packaging - for example, that console apps need to be distributed as `.pkg` bundles on macOS.

It should probably also point out that although you're using argparse in the example code, that's not *required* by Briefcase. You *could* use bare sys.argv handling, click, or any other library you wanted - argparse is just a convenient library that is built into the standard library.

> I have a question thought why is the size of the application on macOS +200MB this is huge for a program that only says Hello + Your Name 😆 I think it would be great if we had a way to make it slimmer.

AFAICT, the *app* isn't 200MB - it's "only" (cough) 89MB. The full build folder is 200MB, but that includes a second copy of the Python framework. 

However, that's still 89MB. So where is that space being used?

About 10MB is libPython; you can't avoid that. A bare bones app with no dependencies adds around 500K - most of which is the app's icon.  You can't avoid that either.

The rest is almost all in the standard library. There's 20MB in the Python test folder, including things such as decimal and compression module test data - but there are also some useful utilities in that folder. There's 21MB in the lib-dynload folder, providing binary modules for the full standard library. There's also all the binary libraries to support Tkinter, OpenSSL, and NCurses. 

However, there's no way for Briefcase to know for sure what your app uses. So - we have to include everything by default.

The good news is that if you know you're not using part of the standard library - you can delete it. Briefcase provides a `cleanup_paths` property that allows you to specify file patterns that you want to be deleted from your packaged app - so if you know you're not going to be using anything in the tests folder or bz2 compression, you can purge those from your app as part of the app release process:
```
cleanup_paths = [
    "My App.app/Contents/Frameworks/Python.framework/Versions/*/lib/python3.*/test",
    "My App.app/Contents/Frameworks/Python.framework/Versions/*/lib/python3.*/lib-dynload/_bz2.*",
]
```
And your app is now 66MB instead of 89MB. Rinse and repeat for anything else you don't need, and you should be able to get a bare bones app down to ~15MB.

FWIW - this would also be worth an explanatory document - either a HOWTO or a tutorial.

john0isaac commented:
I knew you wanted to highlight some points in this one but didn't know what that why I went through and completed it with all the steps based on your other tutorial.

Thanks for the explanation I will try to work on it this weekend it's on of my new year's goals :D

Re size
I agree it should be mentioned somewhere. Have you explored how [renpy](https://www.renpy.org/) does it? They ship installers less than 100 mb with python version included and libs (not sure if it's the full version with tests like you mentioned or not) but it's worth exploring.

freakboy3742 commented:
> They ship installers (full builds) less than 100 mb with python version included and libs (not sure if it's the full version with tests like you mentioned or not) but it's worth exploring.

I'm not sure what you're saying here - the *current* Briefcase installer is less than 100MB - it's 89MB. As I said, the 200MB number you're seeing is the full *build* directory, which includes things that aren't in the distributed final product.

john0isaac commented:
I don't know where did i get this `200 MB` from. I think I got confused.

Sorry for that. To clarify this for anyone seeing this in the future.

The current size for the console application installers are:

- **MacOS .pkg**: `74.1 MB`
- **Windows .msi**: `8.41 MB`
- **Linux .deb**: `62.4 KB`

It's really cool!

freakboy3742 commented:
> I don't know where did i get this `200 MB` from. I think I got confused.

No worries.

> The current size for the console application installers are:
> 
> * **MacOS .pkg**: `74.1 MB`
> * **Windows .msi**: `8.41 MB`
> * **Linux .deb**: `62.4 KB`

To be clear, the 74MB *can* be made *substantially* smaller if you know what parts of the Python standard library you're *not* using. The Linux installer doesn't include any of Python - it's relying on the DEB system package. The Windows installer is also missing a lot of pieces - for example, it doesn't have *any* support for Tkinter, whereas the macOS installer does.

john0isaac commented:
I added all of the points that you mentioned and removed the parts where the command output was exactly the same as the original tutorial referencing that people need to run it with a link to the page.
I left out the main headlines as the steps are the same and can't be skipped. but i merged the build and create in one headline. I hope it's good enough this time.
Let me know if you have anymore comments.

freakboy3742 reviewed (changes requested):
I appreciate the effort you're putting into this PR, and I want to be respectful of your time. However, I'm not sure this is headed in a direction that will ultimately be useful to end users (at least, not without a radical change in direction). 

In my last review, I said there are 5 things this document needs to communicate:

> 1. Explicitly pointing out what is different when you pick "console app";
> 2. Pointing out the console_app flag in pyproject.toml as the major thing that is different
> 3. Highlighting how command line arguments are handled by briefcase dev and briefcase run
> 4. Pointing out notable differences in how console apps operate (e.g., outputting to console, rather than to the system log as GUI apps do)
> 5. Pointing out notable differences in packaging - for example, that console apps need to be distributed as .pkg bundles on macOS.

The current state of the tutorial you've presented here does hit all these 5 points - there's possibly some elaboration needed for points (4) and (5), but the key points are all there. 

The problem is that these 5 salient points (and the importance of them) are buried in a ~500 line document that is mostly a repetition of content that is presented in the original BeeWare tutorial. This tutorial spends more time describing how to use `argparse` than it does describing how Briefcase handles command line arguments - and the latter is the core of what this document should be trying to communicate. 

To be clear - this is almost entirely my fault. Going back over my previous review notes, I'm the one that pointed you in the direction of a tutorial rather a HOWTO. However, looking at where we've ended up, I think that might have been bad advice. I think a much shorter HOWTO document that focusses on the key points I listed above would ultimately be a lot more useful to the end user. That document can assume the user has done the BeeWare tutorial, and has at least a basic familiarity with Briefcase - and can then focus on what is *different* about a CLI app.

I can only apologise for giving you the run-around here. The actual writing you've presented is great - it's just turned out to be the "wrong writing" from the perspective of the gap the we're trying to fill.

I completely understand if I've burned any goodwill I might have had that might prevent you from taking yet another swing at this - again, I can only apologise for providing bad advice that has resulted in wasted effort.

john0isaac commented:
Don't worry about it at all, I learned a lot from the initial feedback about [diataxis](https://diataxis.fr/) as I have never heard of it before and I got to apply it here so that's a win for me. Thanks for sharing it.

I'm happy to give it as many goes as it needs to be acceptable for the beeware documentation section.

I agree with you that the points are between the lines. I will try to highlight them more.
If you don't mind can you share some references or examples to how you want this to be presented?

freakboy3742 commented:
> I agree with you that the points are between the lines. I will try to highlight them more. If you don't mind can you share some references or examples to how you want this to be presented? is this a good example, should i follow it? https://briefcase.readthedocs.io/en/stable/how-to/ci.html

That's about as good an example as any. When writing a HOWTO, you can assume that the audience has basic familiarity with the general workflows and capabilities of the tool - so, they've done the tutorial, and they have a basic idea of the workflows involved in creating and running an app on different platforms. You can then focus on providing instructions on *specific* changes and differences for the topic in question. 

For a console app, I wouldn't be surprised if the 5 bullet points I listed above are literally the section headings for the HOWTO. A light preamble describing the problem the HOWTO is solving; then those 5 points plus some explanations would make a pretty good first draft. There might be some opportunity to collapse some of those points - for example, point 2 is really the "output" of making the change from point 1, so that might just be one heading - but that sort of high level is where you can target the writing.

freakboy3742 commented on docs/how-to/cli-apps.rst:
```diff
@@ -0,0 +1,157 @@
+
+    $ briefcase new
+
+When prompted to select a GUI framework, enter ``4`` to choose **Console**.
```
This won't necessarily be option (4) - it might be a different value depending on the plugins that are installed.

freakboy3742 commented on docs/how-to/cli-apps.rst:
```diff
@@ -0,0 +1,157 @@
+    $ briefcase new
+
+When prompted to select a GUI framework, enter ``4`` to choose **Console**.
+This changes how Briefcase structures the app and configures the project.
```
"structures the app" suggests that the layout of the project will be different - which it won't be. The only thing that changes is the code in `app.py` and the extra config setting.
```suggestion
This changes the code that Briefcase generates for the new app, and the configuration for the new project.
```

freakboy3742 commented on docs/how-to/cli-apps.rst:
```diff
@@ -0,0 +1,157 @@
+
+Briefcase provides two ways to run your app: ``dev`` mode and ``run`` mode.
+You can pass command-line arguments to your app in both cases using ``--`` followed
+by your arguments.
```
This is the right level of detail - but there's a chicken-and-egg issue here. As generated, the output of the app won't say "Hello John" until the argparse stuff is in place.

I'd suggest moving this section to *after* the discussion about handling command line arguments, allowing the example of command line usage to be merged into this section.

freakboy3742 commented on docs/how-to/cli-apps.rst:
```diff
@@ -0,0 +1,157 @@
+    $ briefcase build
+    $ briefcase package
+
+For **macOS**, console apps are packaged as ``.pkg`` files, rather than the ``.app``
+bundles used for GUI apps. This is another major distinction.
```
```suggestion
On **macOS**, console apps *must* be packaged as ``.pkg`` files, rather than the
``.app`` or ``.dmg`` bundles used for GUI apps. This is required to install a
entry on the user's path so that the app can be executed from the command line.
```

freakboy3742 commented on docs/how-to/cli-apps.rst:
```diff
@@ -0,0 +1,157 @@
+    ...
+    [hello-cli] Packaged dist/Hello CLI-0.0.1.pkg
+
+This ``.pkg`` file installs the app globally, and it can run from the terminal:
```
It would be worth noting that the executable name is the app name, not the formal name. This is implied by the `hello-cli` example, but it's worth being explicit (and using `hello-cli` as an example).

freakboy3742 commented on docs/how-to/cli-apps.rst:
```diff
@@ -0,0 +1,157 @@
+Inspecting the File Structure
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Your generated project will have a layout similar to:
```
Only if your app name is `hello-cli` :-) - it would be worth being explicit that this layout is for an app with a formal name of `Hello CLI`, and an app name of `hello-cli`.

freakboy3742 commented on docs/how-to/cli-apps.rst:
```diff
@@ -0,0 +1,157 @@
+Now that you've built a simple console app with Briefcase, you can use these
+principles to build more complex CLI tools or integrate other libraries like
+`Click <https://click.palletsprojects.com/>`_, `Typer <https://typer.tiangolo.com/>`_,
+or any other Python CLI framework.
```
It might also be worth mentioning curses and Textual and here, as a "GUI but in text mode" options.

freakboy3742 reviewed (changes requested):
I've flagged a couple of issues inline; but this looks really good.

freakboy3742 reviewed (approved):
Awesome - thanks for those updates. I've done one final review pass with a bunch of little tweaks; but otherwise this is amazing! Thanks for persisting through the review process (and the dead ends that I sent you down) - the final product was definitely worth the effort.

john0isaac commented:
Thanks!!! It was amazing.
I'm hoping to contribute code changes now (maybe they will be faster :D)
I think the dead ends were a crucial part of the thinking, reasoning, and learning process and I did learn a lot so thank you for your time reviewing and explaining things to me.
